resource-aggregate/events/servicesMetadataSnapshotTaken.go

```diff
@@ -0,0 +1,328 @@
+	offline := make(map[string]*ServicesStatus_Status, len(servicesStatus.GetOnline())+len(servicesStatus.GetOffline()))
+	for _, v := range servicesStatus.GetOffline() {
+		key := v.GetId()
+		offline[key] = v
+	}
+
+	if offline[req.GetStatus().GetId()] != nil {
+		// The service is already offline, and the service needs to be shut down to avoid conflicts in device connection status (ONLINE/OFFLINE).
+		return nil, status.Errorf(codes.FailedPrecondition, "cannot update status for device %v: already offline", req.GetStatus().GetId())
+	}
+
+	online := make(map[string]*ServicesStatus_Status, 1)
+	for _, v := range servicesStatus.GetOnline() {
+		key := v.GetId()
+		if v.GetOnlineValidUntil() < now.UnixNano() {
+			offline[key] = v
+		}
+	}
+	key := req.GetStatus().GetId()
+
+	// When the processing time exceeds the timeToLive, we must extend the timeToLive.
+	// During processing, several tasks are performed: receiving requests, loading data from the database, and processing requests.
+	// Writing to the database is an expensive operation, so if the processing time is more than 10 times the timeToLive, we should extend it.
+	// During idle times, the processing time is typically around 10 milliseconds.
+	// Under heavy load, the processing time can increase to approximately 5 minutes.
+	timeToLive := time.Duration(req.GetStatus().GetTimeToLive())
+	if req.GetStatus().GetTimestamp() >= 0 {
+		processingTime := time.Since(pkgTime.Unix(0, req.GetStatus().GetTimestamp()))
+		if processingTime*writeCostAgainstRead > timeToLive {
+			timeToLive = processingTime * writeCostAgainstRead
+		}
+	}
+
+	online[key] = &ServicesStatus_Status{
+		Id:               req.GetStatus().GetId(),
+		OnlineValidUntil: now.Add(timeToLive).UnixNano(),
+	}
+	delete(offline, key)
+
+	ev := ServicesMetadataUpdated{
+		Status:               &ServicesStatus{Online: serviceStatusMapToArray(online), Offline: serviceStatusMapToArray(offline)},
+		EventMetadata:        em,
+		OpenTelemetryCarrier: propagation.TraceFromCtx(ctx),
+		AuditContext:         ac,
+	}
+	ok, err := d.HandleServicesMetadataUpdated(ctx, &ev)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, nil
+	}
+	return []eventstore.Event{&ev}, nil
+}
+
+// return snapshot as event if any service was confirmed offline
+func (d *ServicesMetadataSnapshotTaken) confirmOfflineServices(ctx context.Context, req *ConfirmOfflineServicesRequest, em *EventMetadata, ac *commands.AuditContext) ([]eventstore.Event, error) {
+	servicesStatus := d.GetServicesMetadataUpdated().GetStatus()
+	offline := make(map[string]*ServicesStatus_Status, len(servicesStatus.GetOffline()))
+	for _, v := range servicesStatus.GetOffline() {
+		key := v.GetId()
+		offline[key] = v
+	}
+
+	removed := make([]*ServicesStatus_Status, 0, len(req.Status))
+	for _, v := range req.Status {
+		key := v.GetId()
+		if _, ok := offline[key]; ok {
+			delete(offline, key)
+			removed = append(removed, v)
+		}
+	}
+	if len(removed) == 0 {
```

Comment:
We can save up a little bit of memory, since we only need to know whether at least one element was removed, you only require a single `bool` for that

Reply:
fixed